controller/controller.c

```diff
@@ -34,56 +34,51 @@ int read_msg(intf_t *intf, char *data, scewl_id_t *src_id, scewl_id_t *tgt_id,
   scewl_hdr_t hdr;
   int read, max;
 
-  do {
-    // clear buffer and header
-    memset(&hdr, 0, sizeof(hdr));
-    memset(data, 0, n);
-
-    // find header start
-    do {
-      hdr.magicC = 0;
+  // clear buffer and header
+  memset(&hdr, 0, sizeof(hdr));
+  memset(data, 0, n);
 
-      if (intf_read(intf, (char *)&hdr.magicS, 1, blocking) == INTF_NO_DATA) {
-        return SCEWL_NO_MSG;
-      }
-
-      // check for SC
-      if (hdr.magicS == 'S') {
-        do {
-          if (intf_read(intf, (char *)&hdr.magicC, 1, blocking) == INTF_NO_DATA) {
-            return SCEWL_NO_MSG;
-          }
-        } while (hdr.magicC == 'S'); // in case of multiple 'S's in a row
-      }
-    } while (hdr.magicS != 'S' && hdr.magicC != 'C');
+  // find header start
+  do {
+    hdr.magicC = 0;
 
-    // read rest of header
-    read = intf_read(intf, (char *)&hdr + 2, sizeof(scewl_hdr_t) - 2, blocking);
-    if(read == INTF_NO_DATA) {
+    if (intf_read(intf, (char *)&hdr.magicS, 1, blocking) == INTF_NO_DATA) {
       return SCEWL_NO_MSG;
     }
 
-    // unpack header
-    *src_id = hdr.src_id;
-    *tgt_id = hdr.tgt_id;
+    // check for SC
+    if (hdr.magicS == 'S') {
+      do {
+        if (intf_read(intf, (char *)&hdr.magicC, 1, blocking) == INTF_NO_DATA) {
+          return SCEWL_NO_MSG;
+        }
+      } while (hdr.magicC == 'S'); // in case of multiple 'S's in a row
+    }
+  } while (hdr.magicS != 'S' && hdr.magicC != 'C');
 
-    // read body
-    max = hdr.len < n ? hdr.len : n;
-    read = intf_read(intf, data, max, blocking);
+  // read rest of header
+  read = intf_read(intf, (char *)&hdr + 2, sizeof(scewl_hdr_t) - 2, blocking);
+  if(read == INTF_NO_DATA) {
+    return SCEWL_NO_MSG;
+  }
 
-    // throw away rest of message if too long
-    for (int i = 0; hdr.len > max && i < hdr.len - max; i++) {
-      intf_readb(intf, 0);
-    }
+  // unpack header
+  *src_id = hdr.src_id;
+  *tgt_id = hdr.tgt_id;
 
-    // report if not blocking and full message not received
-    if(read == INTF_NO_DATA || read < max) {
-      return SCEWL_NO_MSG;
-    }
+  // read body
+  max = hdr.len < n ? hdr.len : n;
+  read = intf_read(intf, data, max, blocking);
 
-  } while (intf != CPU_INTF && intf != SSS_INTF &&                       // always valid if from CPU or SSS
-           ((hdr.tgt_id == SCEWL_BRDCST_ID && hdr.src_id == SCEWL_ID) || // ignore own broadcast
-            (hdr.tgt_id != SCEWL_BRDCST_ID && hdr.tgt_id != SCEWL_ID))); // ignore direct message to other device
+  // throw away rest of message if too long
+  for (int i = 0; hdr.len > max && i < hdr.len - max; i++) {
+    intf_readb(intf, 0);
+  }
+
+  // report if not blocking and full message not received
+  if(read == INTF_NO_DATA || read < max) {
+    return SCEWL_NO_MSG;
+  }
 
```

Comment:
```
if (intf != CPU_INTF && intf != SSS_INTF &&                          // always valid if from CPU or SSS
     ( (hdr.tgt_id == SCEWL_BRDCST_ID && hdr.src_id == SCEWL_ID) ||  // ignore own broadcast
       (hdr.tgt_id != SCEWL_BRDCST_ID && hdr.tgt_id != SCEWL_ID) ))  // ignore direct message to other device
{
return SCEWL_NO_MSG;
}
```

Reply:
In the example code, the `main` function does not check for `read_msg` returning `SCEWL_NO_MSG`. Because the example code doesn't check for `SCEWL_NO_MSG`, I assume that many teams' main functions probably also do not perform this check. Therefore, even if we add an if statement like you suggest, this will still be a breaking change regardless. Whether irrelevant messages are filtered out in `read_msg` or in `main`, teams will still have to modify their current `main` functions, either to check for `SCEWL_NO_MSG` or to perform the filtering. You say that adding the if statement would continue existing functionality, but it really wouldn't. Existing functionality is to never return `SCEWL_NO_MSG` (as long as `blocking == 1`, which is the case in the example).

Therefore, whether to filter out irrelevant messages in `read_msg` or in `main` is just a matter of style, and we (wpi2/Goats team) felt that performing the filtering in `main` is more explicit.